docs: PBTS synchrony issues runbook (#8129)
tendermint/tendermint: docs/tools/debugging/proposer-based-timestamps-runbook.md

---
order: 3
---

# Proposer-Based Timestamps Runbook

Version v0.36 of Tendermint added new constraints for the timestamps included in
each block created by Tendermint. The new constraints mean that validators may
fail to produce valid blocks or may issue `nil` `prevotes` for proposed blocks
depending on the configuration of the validator's local clock.

## What is this document for?

This document provides a set of actionable steps for application developers and
node operators to diagnose and fix issues related to clock synchronization and
configuration of the Proposer-Based Timestamps [SynchronyParams](https://github.com/tendermint/tendermint/blob/841204c9a020cbc255fa117eca2ed6541b3786a7/spec/core/data_structures.md#L450).

## Requirements

To use this runbook, you must be running a node that has the [prometheus metrics endpoint enabled](https://github.com/tendermint/tendermint/blob/841204c9a020cbc255fa117eca2ed6541b3786a7/docs/nodes/metrics.md#L5)
and the Tendermint RPC endpoint enabled and accessible.

It is strongly recommended to also run a Prometheus metrics collector to gather and
analyze metrics from the Tendermint node.

## Debugging a Single Node

> Use this runbook if you observe that your validator is frequently voting `nil` for a block that the rest

Author: Updated, please take a look. I kept some copy in this section as well to make it clear that this is somewhat different from the `Debugging a network` section.

> Use this runbook if you observe that your validator is frequently voting `nil` for a block that the rest

Reviewer: Makes sense to me, thanks!

value to the `SynchronyParams` values for the network. Continue to the [Checking
Sychrony](checking-synchrony) steps.

#### From the `/metrics` url

If you are not running a Prometheus collector, navigate to the `/metrics` endpoint
exposed on the Prometheus metrics port.

Search for the `tendermint_consensus_quorum_prevote_delay` metric. There will be one
entry of this metric for each `proposer_address`. If the value of this metric is
much larger for some proposers, then the issue is likely related to synchronization of their
nodes with NTP. Contact those proposers and ensure that their nodes are properly connected
to NTP using the steps for [debugging a single node](#debugging-a-single-node).

If the values are relatively similar for all proposers you should next compare,
you'll need to compare this value to the `SynchronyParams` for the network. Continue
to the [Checking Sychrony](checking-synchrony) steps.

### Checking Synchrony

To determine the currently configured `SynchronyParams` for your network, issue a
request to your node's RPC endpoint. For a node running locally with the RPC server
exposed on port `26657`, run the following command:

> ```
> # curl localhost:26657/consensus_params
> ```

Reviewer: ```suggestion
```shell
curl localhost:26657/consensus_params
```
```

The json output will contain a field named `synchrony`, with the following structure:

```json
{
  "precision": "500000000",
  "message_delay": "3000000000"
}
```

> The `precision` and `message_delay` values returned are listed in nanoseconds.
> Remember, `tendermint_consensus_quorum_prevote_delay` is listed in seconds.

Reviewer: ```suggestion
The `precision` and `message_delay` values returned are listed in nanoseconds:
In the examples above, the precision is 500ms and the message delay is 3s.
Remember, `tendermint_consensus_quorum_prevote_delay` is listed in seconds.
```

If the `tendermint_consensus_quorum_prevote_delay` value approaches the sum of `precision` and `message_delay`,
then the value selected for these parameters is too small. Your application will

> need to be modified update the `SynchronyParams` to have larger values.

Reviewer: ```suggestion
need to be modified to update the `SynchronyParams` to have larger values.
```

### Updating SynchronyParams

The `SynchronyParams` are `ConsensusParameters` which means they are set and updated
by the application running alongside Tendermint. Updates to these parameters must
be passed to the application during the `FinalizeBlock` ABCI method call.

If the application was built using the CosmosSDK, then these parameters can be updated
programatically using a governance proposal. For more information, see the [CosmosSDK

> documentation]().

Author: My searching has yet to turn up this documentation.

Author: To clarify, the documentation I'm hoping to link here will be just on what a governance proposal is, not necessarily the specific CosmosSDK documentation on PBTS.

If the application does not implement a way to update the consensus parameters

> programatically, then it must be updated to do so. More information on updating

Reviewer: ```suggestion
programatically, then the application itself must be updated to do so. More information on updating
```

This document provides a set of actionable steps for application developers and
node operators to diagnose and fix issues related to clock synchronization and
configuration of the Proposer-Based Timestamps [SynchronyParams](https://github.com/tendermint/tendermint/blob/master/spec/core/data_structures.md#synchronyparams).

Use this runbook if you observe that validators are frequently voting `nil` for a block that the rest
of the network votes for or if validators are frequently producing block proposals
that are not voted for by the rest of the network.

## Requirements

> To use this runbook, you must be running a node that has the [prometheus metrics endpoint enabled](https://github.com/tendermint/tendermint/blob/master/docs/nodes/metrics.md)

Reviewer: ```suggestion
To use this runbook, you must be running a node that has the [Prometheus metrics endpoint enabled](https://github.com/tendermint/tendermint/blob/master/docs/nodes/metrics.md)
```
